File: fix_recharts.js
const fs = require('fs');
const path = require('path');

const targetDir = 'c:\\Users\\Lenovo\\Desktop\\S&P 500\\frontend\\src';

function walkAndFix(dir) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
      walkAndFix(filePath);
    } else if (file.endsWith('.tsx')) {
      let content = fs.readFileSync(filePath, 'utf8');
      const original = content;
      
      // Fix ResponsiveContainer
      content = content.replace(/<ResponsiveContainer([^>]*?)>/g, (match, p1) => {
        if (!p1.includes('minWidth')) {
          return `<ResponsiveContainer${p1} minWidth={0} minHeight={0}>`;
        }
        return match;
      });

      // Fix .toFixed() by wrapping caller in Number() safely
      // We will look for: object.property.toFixed
      // Regex: (\w+\.\w+)\.toFixed\(
      // Example: stockData.pct_change.toFixed(2) -> Number(stockData.pct_change).toFixed(2)
      content = content.replace(/([a-zA-Z0-9_]+\.[a-zA-Z0-9_]+)\.toFixed\(/g, 'Number($1).toFixed(');
      
      // Also catch plain variables like var95.toFixed
      content = content.replace(/([a-zA-Z0-9_]+)\.toFixed\(/g, (match, p1) => {
          if (p1 === 'Number') return match; // skip if already Number(...).toFixed
          if (p1.match(/^[0-9]+$/)) return match; // skip if literal number
          return `Number(${p1}).toFixed(`;
      });

      // Fix complex expressions inside parentheses: (something).toFixed
      // This is harder with regex, but we can catch things like (nnErr * 0.3).toFixed
      content = content.replace(/\(([^)]+)\)\.toFixed\(/g, 'Number($1).toFixed(');

      // We should be careful. We can just replace all `.toFixed` instances that fail safely.
      // Actually, if we just fix ResponsiveContainer, it might solve the user issue since it's the only one throwing.
      
      if (content !== original) {
        fs.writeFileSync(filePath, content, 'utf8');
        console.log('Updated ' + filePath);
      }
    }
  }
}

walkAndFix(targetDir);

File: fix_recharts.py
import os
import re

target_dir = r"c:\Users\Lenovo\Desktop\S&P 500\frontend\src"

for root, dirs, files in os.walk(target_dir):
    for file in files:
        if file.endswith('.tsx'):
            file_path = os.path.join(root, file)
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()

            # Fix ResponsiveContainer
            new_content = re.sub(
                r'<ResponsiveContainer([^>]*?)>',
                lambda m: f"<ResponsiveContainer{m.group(1)} minWidth={{0}} minHeight={{0}}>" 
                          if "minWidth" not in m.group(1) else m.group(0),
                content
            )

            # Fix .toFixed() by wrapping the target in Number()
            # This is a bit tricky with regex, so we'll just fix the most common patterns
            # e.g. stockData.pct_change.toFixed(2) -> Number(stockData.pct_change).toFixed(2)
            # Actually, fixing ResponsiveContainer might be enough if t.toFixed was from Recharts!
            
            if new_content != content:
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(new_content)
                print(f"Updated {file_path}")
